### PR TITLE
Feature/initial with local service calls

### PR DESCRIPTION
**Pull request by aliothe**
Provides the capability to call both local and remote services.
If the service is local, then standard java way is used - if not, tpcall via casual-caller is used.

### CODE REVIEW

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/RemoteRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            if(serviceCallResponse.serviceReturnState() == ServiceReturnState.TPSUCCESS)
+            {
+                CasualBuffer responseBuffer = serviceCallResponse.casualBuffer();
+                String contentType = ContentTypeConverter.convert(CasualBufferType.unmarshall(responseBuffer.getType()));
+                StreamingOutput stream = outputStream -> outputStream.write(responseBuffer.getBytes().getFirst());
+                return Response.ok(stream).header("content-type", contentType).build();
+            }
+            Response.Status status = ErrorStateConverter.convert(serviceCallResponse.errorState());
+            LOG.finest(() -> "service call to " + serviceName + " failed: " + serviceCallResponse.serviceReturnState() + " " + serviceCallResponse.errorState());
+            return Response.status(status).header("content-type", CasualContentType.NULL).build();
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/RemoteRequestHandler.java]: How do we handle scenario where the error occurs with a payload to be returned e.g. service error containing relevant error information in the buffer return.

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        try
+        {
+            List<byte[]> payload = new ArrayList<>();
+            payload.add(IOUtils.toByteArray(inputStream));
+            ServiceBuffer serviceBuffer = ServiceBuffer.of(bufferType.getName(), payload);
+            CasualServiceCallRequestMessage requestMessage =
+                    CasualServiceCallRequestMessage.createBuilder()
+                                                   .setServiceBuffer(serviceBuffer)
+                                                   .setServiceName(serviceName)
+                                                   .setXid(XID.NULL_XID)
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java:58]: Do we need to also add the Flag for NO_TRAN like in remote?

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                                   .build();
+            CasualServiceCallWork work = workCreator.create(UUID.randomUUID(), requestMessage);
+            executorService.runAsync(work).join();
+            CasualServiceCallReplyMessage replyMessage = work.getResponse().getMessage();
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java:62]: We are not checking the errorState of the reply message, which can result in incorrect HTTP response code. Should be similar to remote.

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+            payload.add(IOUtils.toByteArray(inputStream));
+            ServiceBuffer serviceBuffer = ServiceBuffer.of(bufferType.getName(), payload);
+            CasualServiceCallRequestMessage requestMessage =
+                    CasualServiceCallRequestMessage.createBuilder()
+                                                   .setServiceBuffer(serviceBuffer)
+                                                   .setServiceName(serviceName)
+                                                   .setXid(XID.NULL_XID)
+                                                   .build();
+            CasualServiceCallWork work = workCreator.create(UUID.randomUUID(), requestMessage);
+            executorService.runAsync(work).join();
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java:61]: When we `scheduleWork` normally for inbound we utilise the timeout from the network message. Can we use similar here? Or can we should we use a default? What happens if you have a long running operation, which of the configurations "pops" first.

**Review by cklellie — COMMENTED** [review]
Added comments as per our discussion just now on discord.

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java (hunk 4 above)
Comment (aliothe) [reply casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java:61]: We do not have any information re any time out here as it originates from a HTTP POST call.
I'm running tests locally to see what happens when we call a test service that runs for a very long time.

File: versions.gradle
Comment (TobiasLeo) [inline versions.gradle:1]: Maybe we should start using the standardized `libs.versions.toml`-file instead to define dependencies and their version. The nice part is that IntelliJ (and possibly other tools) have a better understanding of such dependencies and can actually resolve the versions to show helpful tips, as well as that it allows easier navigation between usages and the definitions of dependencies https://docs.gradle.org/current/userguide/platforms.html#sub:conventional-dependencies-toml

File: README.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+Calls resulting in anything but TPSUCCESS:
+* ```TPENOENT``` generates 404 ( NOT_FOUND)
+* ```TPETIME``` generates 408 ( REQUEST_TIMEOUT)
+* Any other error generates a 500 ( INTERNAL_SERVER_ERROR) 
+
+These errors are also logged, as finest, with the service name, service return state and error state.
+
+## Exceptional behaviour
+
+Results in a 500 ( INTERNAL_SERVER_ERROR) and the exception is logged as a WARNING.
```
Comment (cklellie) [inline README.md:47]: Do we not return the error in the body response too?

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/CasualService.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+public class CasualService
+{
+    private ServiceCaller serviceCaller;
+    private RemoteRequestHandler remoteRequestHandler;
+    private LocalRequestHandler localRequestHandler;
+    private ExceptionHandler exceptionHandler;
+    private ServiceRegistryLookup serviceRegistryLookup;
+    private CasualServiceCallWorkCreator workCreator;
+    @Resource
+    ManagedExecutorService executorService;
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/CasualService.java]: This is not used, remove.

File: casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import se.laz.casual.network.protocol.messages.service.CasualServiceCallRequestMessage;
+
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+public class LocalRequestHandler
+{
+    @Resource
```
Comment (cklellie) [inline casual/casual-http/src/main/java/se/laz/casual/http/resources/handlers/LocalRequestHandler.java]: Use a producer to allow this to be injected through the constructor.
Will remove the need to use groovy to access the private member variable in the test too.

File: casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+      serviceExists(_) >> true
+   }
+
+   @Unroll
+   def 'remote service call ok #mimeType #expectedReturnMimeType'()
+   {
+      given:
+      Dispatcher dispatcher = MockDispatcherFactory.createDispatcher()
+      ServiceCaller serviceCaller = Mock(ServiceCaller){
+         1 * makeServiceCall(_, serviceName, Flag.of(AtmiFlags.TPNOTRAN)) >> { CasualBuffer msg, String serviceName, Flag<AtmiFlags> flags ->
```
Comment (cklellie) [inline casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy:68]: Wildcard without checking the contents of the payload. Won't catch incorrect calls to makeServiceCall.

File: casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+            new ServiceCallResponse(ServiceReturnState.TPSUCCESS, ErrorState.OK, replyBuffer)
+         }
+      }
+      CasualService casualService = new CasualService(serviceCaller, new RemoteRequestHandler(), Mock(LocalRequestHandler), Mock(ExceptionHandler), serviceRegistryLookupServiceDoesNotExist)
+      dispatcher.getRegistry().addSingletonResource(casualService)
+      MockHttpRequest request = MockHttpRequest.post("${root}/${serviceName}")
+              .contentType(mimeType)
+              .content(requestBuffer.getBytes().first)
+      MockHttpResponse response = new MockHttpResponse()
+      expect:
```
Comment (cklellie) [inline casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy]: expect here feels wrong. It should be when then rather as you are doing something within the test.
expect is more for a setup assertion check.
here you perform an http call and then check the response rather.

File: casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+      dispatcher.invoke(request, response)
+      response.getStatus() == Response.Status.OK.statusCode
+      response.outputHeaders[HttpHeaders.CONTENT_TYPE].first.toString().contains(expectedReturnMimeType)
+      CasualBuffer responseBuffer = creatorFunction(response.output)
+      responseBuffer.getBytes() == requestBuffer.getBytes()
```
Comment (cklellie) [inline casual/casual-http/src/test/groovy/se/laz/casual/http/resources/CasualServiceTest.groovy:87]: no check that the local service implementation was not called during a remote call.

File: publish.gradle [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+            versionMapping {
+                usage( 'java-api' ) {
+                    fromResolutionOf( 'runtimeClasspath' )
+                }
+                usage( 'java-runtime' ) {
+                    fromResolutionResult()
+                }
+            }
+            pom {
+                name = 'Casual Caller'
```
Comment (cklellie) [inline publish.gradle]: Copy pasta, fail.

**Review by cklellie — COMMENTED** [review]
Generally looks good. Added some comments for a few smaller things.
I do though find the main unit test for the rest layer is hard to follow, it could do with refactoring and make more valuable assertions to ensure correctness as well as making the assertions clearer.
I have added some comments on one the test cases, but they apply to each of the remaining test cases.

File: README.md (hunk 5 above)
Comment (aliothe) [reply README.md:47]: No we do not, we log it but we do not return it

File: versions.gradle
Comment (aliothe) [reply versions.gradle:1]: @cklellie what do you think?
Does the current versions.gradle thing not work nicely with IntelliJ?

Comment (aliothe) [reply versions.gradle:1]: Sounds interesting, we could look at it going forward but not at this moment.
